src/tiledb/cloud/geospatial/ingestion.py

```diff
@@ -0,0 +1,1287 @@
+    :param tile_size: for rasters this is the tile (block) size
+        for the merged destination array, defaults to 1024
+    :param chunk_size: for point cloud this is the PDAL chunk size, defaults to 1000000
+    :param nodata: NODATA value for raster merging
+    :param resampling: string, resampling method,
+        one of None, bilinear, cubic, nearest and average
+    :param res: Tuple[float, float], output resolution in x/y
+    :param verbose: verbose logging, defaults to False
+    :param trace: bool, enabling log tracing, defaults to False
+    :param log_uri: log array URI
+    """
+
+    logger = get_logger_wrapper(verbose)
+
+    if log_uri:
+        create_log_array(log_uri)
+
+    kwargs = {}
+    fn = None
+
+    if dataset_list_uri:
+        sources = read_uris(
+            dataset_list_uri,
+            dataset_type=dataset_type,
+            log_uri=log_uri,
+            config=config,
+            max_files=max_files,
+        )
+
+    def pointcloud_match(f):
+        return os.path.splitext(f)[-1].lstrip(".").lower() in ["las", "laz"]
+
+    def raster_match(f):
+        try:
+            rasterio.driver_from_extension(f)
+            return True
+        except ValueError:
+            return False
+
+    def geometry_match(f):
+        try:
+            fiona.drvsupport.driver_from_extension(f)
+            return True
+        except ValueError:
+            return False
+
+    funcs = {
+        DatasetType.POINTCLOUD: {
+            "udf_fn": ingest_point_cloud_udf,
+            "pattern_fn": pointcloud_match,
+            "meta_fn": get_pointcloud_metadata,
+        },
+        DatasetType.RASTER: {
+            "udf_fn": ingest_raster_udf,
+            "pattern_fn": raster_match,
+            "meta_fn": get_raster_metadata,
+        },
+        DatasetType.GEOMETRY: {
+            "udf_fn": ingest_geometry_udf,
+            "geometry_fn": geometry_match,
+            "meta_fn": get_geometry_metadata,
+        },
+    }
+
+    if search_uri:
+        sources = find(
+            search_uri,
+            config=config,
+            excludes=ignore,
+            includes=pattern if pattern else funcs[dataset_type]["pattern_fn"],
+            max_files=max_files,
+        )
+
+    if len(sources) == 0:
+        raise ValueError(f"No {dataset_type.name} datasets found")
+
+    meta = funcs[dataset_type]["meta_fn"](sources, config=config)
+
+    if dataset_type == DatasetType.POINTCLOUD:
+        if len(meta.paths) > 0:
+            kwargs.update(
+                {
+                    "template_sample": meta.paths[0],
+                    "extents": meta.extents,
+                    "offsets": meta.offsets,
+                    "scales": meta.scales,
+                    "chunk_size": chunk_size,
+                }
+            )
+            samples = meta.paths[1:]
+        else:
+            raise ValueError("Require at least one point cloud file to have been found")
```

Comment:
insetad of

```python
if something:
  do_stuff()
else:
  raise ...
```

This can be written as
```python
if not something:
  raise ...
do_stuff()
```
to reduce the amount of nesting and make the "happy path" easier to follow.

Reply:
👍